### PR TITLE
Configuration: Enable support for memory pre-allocation and swap. Also enable support for debug mode.

### PR DESCRIPTION
**Pull request by mcastelino**
Depends on vendoring of QEMU https://github.com/01org/ciao/pull/1434

# Add support for fine grained control of memory management.

    ```enable_prealloc=true```  Will result in the upfront allocation  and locking down of all VM memory, resulting in lower VM density.

    ```enable_swap=true```   Will enable swap of VM memory, resulting in higher VM density with potentially higher latency

Enabling the two together results in undefined behavior.

    ```enable_debug=true```   Will enable hypervisor and container kernel logs

Fixes https://github.com/clearcontainers/runtime/issues/568

### CODE REVIEW

File: config/configuration.toml.in [hunk 1]
```diff
@@ -20,6 +20,9 @@ default_vcpus = -1
 #default_memory = @DEFMEMSZ@
 disable_block_device_use = @DEFDISABLEBLOCK@
 
+# Enable pre allocation of VM RAM, default false
```
Comment (sameo) [inline config/configuration.toml.in]: We probably want to explain why we'd want to set that one to true and what would be the consequences as far as memory footprint would be (i.e. bigger).

**Comment by mcastelino** [comment]
Please re-review. Added more knobs which should help make the whole feature fully configurable.

File: config/configuration.toml.in [hunk 2]
```diff
@@ -20,6 +20,17 @@ default_vcpus = -1
...
+#enable_mem_prealloc = true
+#
+# Enable Realtime QEMU
+#enable_realtime = true
```
Comment (jodh-intel) [inline config/configuration.toml.in]: Similarly, I think we should document why this isn't the default - can you document in the comments the implication of enabling this?

File: config/configuration.toml.in [hunk 3]
```diff
@@ -20,6 +20,17 @@ default_vcpus = -1
...
+#
+# Enable Mlock
```
Comment (jodh-intel) [inline config/configuration.toml.in]: Can you expand the title further. Could this be `Enable memory locking` for example?

Also, `qemu(1)` only seems to provide a single option that handles both realtime and mlock, so why do we need (and how are we providing) two knobs?

**Comment by grahamwhaley** [comment]
Agree that we could do with comments (similar to the ones you have in the commits) in probably the toml file.
CIs failed - mostly I expect as we'll need the re-vendor/rebase to pick up the necessary virtcontainers bits? - but, also, I think you have no `# fixes` line in a commit to satisfy checkcommits.

File: config/configuration.toml.in (hunk 3 above)
Comment (mcastelino) [reply config/configuration.toml.in]: By default memory once allocated is always locked. Unfortunately to be able to swap memory in qemu the realtime option must be enabled. 

So our default behavior is locked non pre-allocated. So memory is allocated on behalf of the VM, when the memory is actually touched by the VM. Once memory is brought in, it is not swapped out.

So our behaviour is

1. Default, non pre-allocated but locked (with none of the knobs set)
2. Realtime + Locked
    Which gives you realtime behaviour
3. Realtime + Unlocked (i.e. swap enabled)
    Here what we desire is swap enabled, but we have to enable realtime which is unfortunate.
    But once you enable swap, it is not realtime anyway. So effectively it is just enabling swap.
    So in the implementation, we can choose to honor the mlock=off w/o requiring the realtime 
    knob to be set.
   
4. You can also enable pre-allocated, realtime and locked (which does not work today btw. The agent seems to be failing. 
   Note: This is not an issue for normal VM's. It just triggers some error in Clear Containers, the agent in particular.

There will be bug filed for 4, once this PR is merged to track that issue.

File: config/configuration.toml.in (hunk 2 above)
Comment (mcastelino) [reply config/configuration.toml.in]: We can make this the default post PnP to evaluate its impact.

**Comment by mcastelino** [comment]
Based on all the feedback here I am torn between keeping the knobs raw and infinitely configurable or exposing abstract tuneables. Say
```
- enable_prealloc - Default false
- enable_realtime - Default false (for now) 
- enable_swap - Default false
```

In that case however I would add another open ended param say
```
-extra_qemu_args 
```
which takes a free form set of params, as this will be needed when you need infinite configuration.

**Comment by mcastelino** [comment]
Updated the PR to make things a little more implicit.  However QEMU needs to be changed yet again to get desired behavior.
https://github.com/01org/ciao/pull/1434

**Comment by grahamwhaley** [comment]
Travis/checkcommits notes you are missing a 'Fixes' in your commits.

File: config/configuration.toml.in [hunk 4]
```diff
@@ -20,6 +20,24 @@ default_vcpus = -1
...
 
+# Enable pre allocation of VM RAM, default false
+# Enabling this will result in lower container density
+# as all of the memory will be allocated and locked
+# This is useful when you want to reserve all the memory
+# upfront or in the cases where you want memory latencies
+# to be very predictable
+# Default false
+#enable_mem_prealloc = true
+#
```
Comment (grahamwhaley) [inline config/configuration.toml.in]: Replace these `# ` blank lines with just blank lines to keep in the style of the rest of the file.
Although, the reset of the file is not wholly consistent, I think breaking up these sections will fully blank lines will make it clearer.

File: config/configuration.toml.in [hunk 5]
```diff
@@ -20,6 +20,24 @@ default_vcpus = -1
...
+# Enable swap of vm memory. Default false.
+# The behaviour is undefined if mem_prealloc is also set to true
```
Comment (grahamwhaley) [inline config/configuration.toml.in]: Note here (the sort of inverse of the pre-alloc) - that this will likely increase density at the cost of latency.

File: Makefile [hunk 6]
```diff
@@ -111,6 +111,7 @@ DEFMEMSZ := 2048
 DEFDISABLEBLOCK := false
 DEFENABLEMEMPREALLOC := false
 DEFENABLESWAP := false
+DEFENABLEDEBUG := false
```
Comment (grahamwhaley) [inline Makefile]: My brain initially thought this should be a separate PR, but on reconsideration it's fine - all three commits modify the config - just happens that two of the commits appear related :-)

File: config/configuration.toml.in [hunk 7]
```diff
@@ -20,6 +20,24 @@ default_vcpus = -1
...
+#enable_swap = true
+#
+# Debug changes the default hypervisor and kernel parameters to
+# enable debug output where available.
```
Comment (jodh-intel) [inline config/configuration.toml.in]: Can you put a comment explaining where that debug output goes / how to view it?

Comment (mcastelino) [reply config/configuration.toml.in]: Will do. I am also writing an Advanced debugging readme. I will add that to it too.

That will consist of
https://github.com/clearcontainers/runtime/issues/559
https://github.com/clearcontainers/runtime/issues/558

and anything else appropriate

**Comment by mcastelino** [comment]
@grahamwhaley @jodh-intel addressed your review comments.

Note: We need to re-vendor ciao qemu.

**Comment by sameo** [comment]
Depends on PR #580

**Comment by sameo** [comment]
PR #580 merged
